sources/drbd_resource_repair.py

```diff
@@ -0,0 +1,292 @@
+import argparse
+import json
+import logging
+import os
+import shlex
+import subprocess
+from functools import lru_cache
+
+from typing import Any, Dict, Iterator, List, Optional, Set, Union
+
+SCRIPT_NAME = "drbd_resource_repair"
+DRBD_STATUS = "drbdsetup status all --json"
+DRBD_VERIFY = "drbdadm verify {resource}:{peer}/0"
+DRBD_WAIT_SYNC = "drbdadm wait-sync {resource}"
+DRBD_INVALIDATE = "drbdadm invalidate {resource}:{peer}/0 --reset-bitmap=no"
+DRBD_INVALIDATE_REMOTE = (
+    "drbdadm invalidate-remote {resource}:{peer}/0 --reset-bitmap=no"
+)
+
+
+def run_command(cmd: str, ignore_dry_run: bool = False, remote_host: str = "") -> str:
+    if remote_host:
+        cmd = f"ssh root@{remote_host} -C {cmd}"
+    logger.debug(cmd)
+    if DRY_RUN and not ignore_dry_run:
+        return ""
+    try:
+        return subprocess.run(
+            shlex.split(cmd),
+            check=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+        ).stdout.strip()
+    except subprocess.CalledProcessError as e:
+        logger.error("%s failed with error code %d: `%s`", cmd, e.returncode, e.output)
+        raise e
+
+
+@lru_cache()
+def get_hostname(ip: str) -> str:
+    return run_command("hostname", ignore_dry_run=True, remote_host=ip)
+
+
+class Host:
+    def __init__(self, ip: str):
+        self.ip = ip
+
+    @property
+    def hostname(self) -> str:
+        return get_hostname(self.ip)
+
+    def __str__(self) -> str:
+        return ",".join([self.hostname, self.ip])
+
+    def __eq__(self, other: Any) -> bool:
+        return (
+            isinstance(other, Host)
+            and self.ip == other.ip
+            and self.hostname == other.hostname
+        )
+
+    def __hash__(self) -> int:
+        return hash((self.ip, self.hostname))
+
+    def to_json(self) -> Dict[str, Any]:
+        return self.__dict__
+
+class LinstorPeer(Host):
+    def __init__(self, ip: str, from_ip: str):
+        super().__init__(ip)
+        self.from_host = Host(from_ip)
+
+    def __str__(self) -> str:
+        return ",".join([super().__str__(), str(self.from_host)])
+
+    # Intentionally not overloading eq and hash since the from_host
+    #  does not change the peer
+
+
+class ResourceStatus:
+    def __init__(self, resource: str, peer: LinstorPeer, out_of_sync: int):
+        self.resource = resource
+        self.peer = peer
+        self.out_of_sync = out_of_sync
+
+    def __str__(self) -> str:
+        return ",".join([self.resource, str(self.peer), str(self.out_of_sync)])
+
+    def __eq__(self, other: Any) -> bool:
+        return (
+            isinstance(other, ResourceStatus)
+            and self.resource == other.resource
+            and self.peer == other.peer
+            and self.out_of_sync == other.out_of_sync
```

Comment:
I think you can remove the last comparison, the key of a resource status should be a resource name and a peer name. In fact you can even have a `ResourceStatusKey` or something like that to split the key and the associated data.